### ISSUE TITLE
Trouble typing object whose values are of a particular type while maintaining the names of its keys

### ISSUE THREAD

**Issue by rigdern** [Microsoft/TypeScript#5917]
I'm trying to create an object and I want type checking to work in two different scenarios but I can't get both scenarios to pass the type checker at the same time.

I have an object that I'm essentially treating as a set of key-value pairs. Each of these values is of the same type. Here are the two scenarios I want to achieve with this object:
  1. When accessing properties on the object, I want the type checker to be aware of the particular keys on the object that have been defined.
  2. I want to be able to pass the object to a function which accepts any object whose value are all of a particular type (it doesn't care about the names of the keys).

Here's a complete example:

```ts
class Person {
	constructor(public age: number) {
	}
}

// If you use the commented out version of people, Goal 2 is achieved
// but Goal 1 breaks.
//var people: { [name: string]: Person } = {
var people = {
	john: new Person(10),
	pete: new Person(20),
	tim: new Person(30),
	sam: new Person(40),
};

// Goal 1: I want to get intellisense on people
var somePerson = people.john;

// Goal 2: I want to have a function which accepts an object whose values are all
// of type Person.
function inspectPeople(people: { [name: string]: Person }) {
}
inspectPeople(people);
```

Here's a link to [this example in the TypeScript Playground](http://www.typescriptlang.org/Playground#src=class%20Person%20%7B%0A%09constructor(public%20age%3A%20number)%20%7B%0A%09%7D%0A%7D%0A%0A%2F%2F%20If%20you%20use%20the%20commented%20out%20version%20of%20people%2C%20Goal%202%20is%20achieved%0A%2F%2F%20but%20Goal%201%20breaks.%0A%2F%2Fvar%20people%3A%20%7B%20%5Bname%3A%20string%5D%3A%20Person%20%7D%20%3D%20%7B%0Avar%20people%20%3D%20%7B%0A%09john%3A%20new%20Person(10)%2C%0A%09pete%3A%20new%20Person(20)%2C%0A%09tim%3A%20new%20Person(30)%2C%0A%09sam%3A%20new%20Person(40)%2C%0A%7D%3B%0A%0A%2F%2F%20Goal%201%3A%20I%20want%20to%20get%20intellisense%20on%20people%0Avar%20somePerson%20%3D%20people.john%3B%0A%0A%2F%2F%20Goal%202%3A%20I%20want%20to%20have%20a%20function%20which%20accepts%20an%20object%20whose%20values%20are%20all%0A%2F%2F%20of%20type%20Person.%0Afunction%20inspectPeople(people%3A%20%7B%20%5Bname%3A%20string%5D%3A%20Person%20%7D)%20%7B%0A%7D%0AinspectPeople(people)%3B%0A).

**Comment by mhegazy** [comment]
And you do not want to specify the "known" properties on the type?

```ts
var people: { [name: string]: Person; john: Person, pete:Person, tim:Person, sam:Person} = {
	john: new Person(10),
	pete: new Person(20),
	tim: new Person(30),
	sam: new Person(40),
};
```

**Comment by rigdern** [comment]
Thanks for the suggestion! I was not aware of this solution and it does fix the type errors.

However, I wish there was a less verbose/repetitive solution where something closer to my original definition of `people` would work:

```ts
var people = {
    john: new Person(10),
    pete: new Person(20),
    tim: new Person(30),
    sam: new Person(40),
};
```

Here are some more concise solutions I had in mind that don't seem to currently exist in TypeScript:
  1. Allow the `{ [name: string]: Person }` signature of `inspectPeople` to accept an object like `people`.
  2. Instead of changing the meaning of  `{ [name: string]: Person }`, introduce an alternative notation for expressing a type that would match any object whose properties are all of type `Person` (e.g. like `people`).
  3. Allow developers to specify that, in addition to the key-value pairs defined, `people` supports an indexer so it can be passed to a function like `inpsectPeople`.

If there's no more concise way to express this scenario in the current version of TypeScript, are there thoughts on how it might be expressed in future versions or is the current solution considered to be good enough?

**Comment by ahejlsberg** [comment]
You could do the following:

```typescript
function makePeople<T extends { [name: string]: Person }>(people: T): T {
    return people;
}

var people = makePeople({
    john: new Person(10),
    pete: new Person(20),
    tim: new Person(30),
    sam: new Person(40)
});
```

This will give `people` an inferred type that includes both the indexer and the explicitly specified members:

```typescript
var people: {
    [name: string]: Person;
    john: Person;
    pete:Person;
    tim:Person;
    sam:Person;
};
```

Plus it will verify that the argument passed to `makePeople` is indeed an object that contains only properties of type `People`.

**Comment by rigdern** [comment]
I like that solution. Thanks!

**Comment by litera** [comment]
What if we wanted to have additional properties that are not of type `Person`? How do we do that?
```typescript
interface IPeople {
    [name: string]: Person;
    count: number;
}
```
This gives an error of
> Property 'count' of type 'number' is not assignable to string index type 'Person'.

**Comment by RyanCavanaugh** [comment]
@litera that is correct. Observe:
```ts
let x: string = "cou" + "nt";
let a: IPeople = ...;
let p = a[x]; // typeof p === 'number', but we would think it was 'Person'
```

**Comment by litera** [comment]
@RyanCavanaugh That can be yes, but what if you'd want to add many properties of various types? That would just make it a mess, don't you think?

**Comment by kitsonk** [comment]
> That would just make it a mess, don't you think?

But how to propose to properly represent the index access then?  Messy or not.

**Comment by RyanCavanaugh** [comment]
An object that is both a bunch of declared properties *and* a lookup table on arbitrary keys does sound like a mess.

**Comment by kitsonk** [comment]
> An object that is both a bunch of declared properties and a lookup table on arbitrary keys does sound like a mess.

Agreed.  On the other hand, some people don't choose that as self inflected pain.  I ran across a situation the other day that I tried to model in TypeScript, a POJO DSL (Mongo DB) where objects combined arbitrary field names with "special" properties that had a different valid type.  It is still an unsafe mess because of this in TypeScript, but it is still better than just praying to the gods of the sever to properly parse the POJO.
